File: server/src/fieldParsers.js
'use strict';

const DATE_SANITY_CONFIG = {
  minYear: 2020,
  // "current_year+1" per spec — computed at call time, not baked in.
};

/** Sq.Ft. / Acres: null (not 0) when unparseable, e.g. "N/A", "None - MPAC". */
function parseNumericField(raw) {
  if (raw === null || raw === undefined) return { value: null, raw: null };
  if (typeof raw === 'number') return { value: raw, raw: String(raw) };
  const s = String(raw).trim();
  if (s === '') return { value: null, raw: s };
  const cleaned = s.replace(/,/g, '');
  if (/^-?\d+(\.\d+)?$/.test(cleaned)) {
    return { value: parseFloat(cleaned), raw: s };
  }
  return { value: null, raw: s }; // "N/A", "None - MPAC", etc. — keep raw, no guessing
}

/**
 * Building Coverage % — source files store this as a fraction (0.0982...) with
 * a percent number format, or occasionally as literal text like "18%"/"N/A".
 */
function parseCoveragePct(raw) {
  if (raw === null || raw === undefined) return null;
  if (typeof raw === 'number') return raw <= 1 ? raw : raw / 100;
  const s = String(raw).trim();
  const m = s.match(/^(-?\d+(\.\d+)?)\s*%$/);
  if (m) return parseFloat(m[1]) / 100;
  return null;
}

/**
 * Taxes/TMI: store raw as-is always; extract a numeric $ value only when the
 * pattern is confidently parseable ("$52,500 Annual", "$9,900/Yr", plain "5.5").
 * Narrative strings ("50% of Hydro Bill") and "N/A" are left null, never guessed.
 */
function parseTaxesNumeric(raw) {
  if (raw === null || raw === undefined) return null;
  const s = String(raw).trim();
  if (s === '' || /^n\/?a$/i.test(s)) return null;
  if (/%/.test(s) && /of\s+\w+/i.test(s)) return null; // "50% of Hydro Bill" — narrative, don't guess
  const m = s.match(/\$?\s*([\d,]+(?:\.\d+)?)/);
  if (!m) return null;
  const n = parseFloat(m[1].replace(/,/g, ''));
  return Number.isNaN(n) ? null : n;
}

/**
 * Date parsing + sanity check. Accepts a JS Date (from SheetJS cellDates:true),
 * an Excel serial number, or a string. Returns { iso, raw, flagged } — flagged
 * is true when the parsed year falls outside [minYear, currentYear+1], per spec
 * ("typo'd year 2206 instead of 2026" must be caught, not silently kept/discarded).
 */
function parseDateField(raw) {
  const rawStr = raw instanceof Date ? raw.toISOString() : String(raw ?? '');
  const nowYear = new Date().getFullYear();
  const maxYear = nowYear + 1;

  let d = null;
  if (raw instanceof Date && !Number.isNaN(raw.getTime())) {
    d = raw;
  } else if (typeof raw === 'number') {
    // Excel serial date (days since 1899-12-30)
    d = new Date(Date.UTC(1899, 11, 30) + raw * 86400000);
  } else if (typeof raw === 'string' && raw.trim() !== '') {
    const parsed = new Date(raw);
    if (!Number.isNaN(parsed.getTime())) d = parsed;
  }

  if (!d || Number.isNaN(d.getTime())) {
    return { iso: null, raw: rawStr, flagged: false };
  }

  const year = d.getUTCFullYear();
  const flagged = year < DATE_SANITY_CONFIG.minYear || year > maxYear;
  const iso = d.toISOString().slice(0, 10);
  return { iso, raw: rawStr, flagged };
}

/** Normalize vendor name for grouping (casing/whitespace) while keeping raw. */
function normalizeVendor(raw) {
  if (raw === null || raw === undefined) return { raw: null, normalized: null };
  const s = String(raw).replace(/ /g, ' ').trim(); // strip stray non-breaking spaces (seen in source files)
  if (s === '') return { raw: s, normalized: null };
  if (/^unknown$/i.test(s)) return { raw: s, normalized: 'Unknown' };
  const normalized = s
    .replace(/\s+/g, ' ')
    .replace(/[.,]+$/, '')
    .toUpperCase();
  return { raw: s, normalized };
}

module.exports = {
  parseNumericField,
  parseCoveragePct,
  parseTaxesNumeric,
  parseDateField,
  normalizeVendor,
  DATE_SANITY_CONFIG,
};

File: streamlit_app/lib/field_parsers.py
"""Dates, sq ft/acres, taxes, vendor name normalization — ported from
fieldParsers.js.
"""

import re
import datetime as dt

DATE_SANITY_MIN_YEAR = 2020


def parse_numeric_field(raw):
    """Sq.Ft. / Acres: None (not 0) when unparseable, e.g. 'N/A', 'None - MPAC'."""
    if raw is None:
        return {"value": None, "raw": None}
    if isinstance(raw, (int, float)):
        return {"value": float(raw), "raw": str(raw)}
    s = str(raw).strip()
    if s == "":
        return {"value": None, "raw": s}
    cleaned = s.replace(",", "")
    if re.match(r"^-?\d+(\.\d+)?$", cleaned):
        return {"value": float(cleaned), "raw": s}
    return {"value": None, "raw": s}


def parse_coverage_pct(raw):
    if raw is None:
        return None
    if isinstance(raw, (int, float)):
        return float(raw) if raw <= 1 else float(raw) / 100
    s = str(raw).strip()
    m = re.match(r"^(-?\d+(\.\d+)?)\s*%$", s)
    if m:
        return float(m.group(1)) / 100
    return None


def parse_taxes_numeric(raw):
    if raw is None:
        return None
    s = str(raw).strip()
    if s == "" or re.match(r"^n/?a$", s, re.I):
        return None
    if "%" in s and re.search(r"of\s+\w+", s, re.I):
        return None  # "50% of Hydro Bill" — narrative, don't guess
    m = re.search(r"\$?\s*([\d,]+(?:\.\d+)?)", s)
    if not m:
        return None
    try:
        return float(m.group(1).replace(",", ""))
    except ValueError:
        return None


def parse_date_field(raw):
    """Returns {iso, raw, flagged}. flagged=True when the parsed year is
    outside [MIN_YEAR, current_year+1] — e.g. the known '2206' instead of
    '2026' typo — so it's caught, not silently kept or discarded.
    """
    now_year = dt.datetime.now().year
    max_year = now_year + 1

    d = None
    raw_str = ""
    if isinstance(raw, dt.datetime):
        d = raw
        raw_str = raw.isoformat()
    elif isinstance(raw, dt.date):
        d = dt.datetime(raw.year, raw.month, raw.day)
        raw_str = raw.isoformat()
    elif isinstance(raw, (int, float)):
        # Excel serial date (days since 1899-12-30)
        raw_str = str(raw)
        try:
            d = dt.datetime(1899, 12, 30) + dt.timedelta(days=raw)
        except (OverflowError, ValueError):
            d = None
    elif isinstance(raw, str) and raw.strip() != "":
        raw_str = raw
        for fmt in ("%Y-%m-%d", "%Y-%m-%dT%H:%M:%S", "%m/%d/%Y", "%m/%d/%y", "%Y-%m-%d %H:%M:%S"):
            try:
                d = dt.datetime.strptime(raw.strip(), fmt)
                break
            except ValueError:
                continue
    elif raw is None:
        raw_str = ""

    if d is None:
        return {"iso": None, "raw": raw_str, "flagged": False}

    year = d.year
    flagged = year < DATE_SANITY_MIN_YEAR or year > max_year
    return {"iso": d.strftime("%Y-%m-%d"), "raw": raw_str, "flagged": flagged}


def normalize_vendor(raw):
    if raw is None:
        return {"raw": None, "normalized": None}
    s = str(raw).replace("\xa0", " ").strip()
    if s == "":
        return {"raw": s, "normalized": None}
    if re.match(r"^unknown$", s, re.I):
        return {"raw": s, "normalized": "Unknown"}
    normalized = re.sub(r"\s+", " ", s)
    normalized = re.sub(r"[.,]+$", "", normalized).upper()
    return {"raw": s, "normalized": normalized}
